Shell/InductionSchemeGenerator.hpp

```diff
@@ -0,0 +1,243 @@
+/*
+ * This file is part of the source code of the software program
+ * Vampire. It is protected by applicable
+ * copyright laws.
+ *
+ * This source code is distributed under the licence found here
+ * https://vprover.github.io/license.html
+ * and in the source directory
+ */
+/**
+ * @file InductionSchemeGenerator.hpp
+ * Defines classes for generating induction schemes from function terms
+ */
+
+#ifndef __InductionSchemeGenerator__
+#define __InductionSchemeGenerator__
+
+#include "Forwards.hpp"
+#include "Kernel/Substitution.hpp"
+#include "Kernel/Term.hpp"
+#include "Kernel/TermTransformer.hpp"
+#include "InductionPreprocessor.hpp"
+#include "Lib/STL.hpp"
+
+#include <bitset>
+
+namespace Shell {
+
+using namespace Kernel;
+using namespace Lib;
+using namespace Indexing;
+
+/* Provides a wrapper for a bit vector. After setting its bits and size, it
+ * can iterate through all combinations of bit vectors masking its 1 bits. */
+class Occurrences {
+public:
+  Occurrences(bool initial)
+    : _occ(1 & initial), _iter(), _max(1 << 1), _finished(false) {}
+
+  void add(bool val) {
+    ASS(!_finished);
+    _occ <<= 1;
+    _max <<= 1;
+    _occ |= (1 & val);
+  }
+
+  // reverse the bit vector
+  void finalize() {
+    ASS(!_finished);
+    _finished = true;
+    const auto c = num_bits();
+    ASS(c); // disallow empty bitvectors
+    auto temp = _occ;
+    _occ = 0;
+    for (uint64_t i = 0; i < c; i++) {
+      _occ <<= 1;
+      _occ |= temp & 1;
+      temp >>= 1;
+    }
+    _iter = _occ;
+  }
+
+  // iterate to the next bit vector
+  bool next() {
+    ASS(_finished);
+    _iter++;
+    _iter |= _occ; // mask initial bits
+    return _iter < _max;
+  }
+
+  bool pop_last() {
+    ASS(_finished);
+    bool res = _iter & 1;
+    _iter >>= 1;
+    _max >>= 1;
+    return res;
+  }
+
+  uint64_t val() {
+    ASS(_finished);
+    return _iter;
+  }
+
+  uint64_t num_bits() const {
+    ASS(_finished);
+    return __builtin_ctz(_max);
```

Comment:
Maybe add here ASS(_max);? Documentation for __builtin_ctz(x) says that the result is undefined if x==0. https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html

Reply:
Good catch! I added the assertion.